modules/videoio/src/cap_libcamera.hpp

```diff
@@ -0,0 +1,539 @@
+                                unsigned int *stride = nullptr) const;
+    Stream *StillStream(unsigned int *w = nullptr, unsigned int *h = nullptr, unsigned int *stride = nullptr) const;
+    Stream *RawStream(unsigned int *w = nullptr, unsigned int *h = nullptr, unsigned int *stride = nullptr) const;
+    Stream *VideoStream(unsigned int *w = nullptr, unsigned int *h = nullptr, unsigned int *stride = nullptr) const;
+    Stream *LoresStream(unsigned int *w = nullptr, unsigned int *h = nullptr, unsigned int *stride = nullptr) const;
+    Stream *GetMainStream() const;
+
+    std::vector<libcamera::Span<uint8_t>> Mmap(FrameBuffer *buffer) const;
+
+    void SetControls(ControlList &controls);
+    void StreamDimensions(Stream const *stream, unsigned int *w, unsigned int *h, unsigned int *stride) const;
+
+protected:
+    std::unique_ptr<Options> options_;
+
+private:
+    static std::shared_ptr<CameraManager> getCameraManager()
+    {
+        static std::shared_ptr<CameraManager> camera_manager_;
+        if (!camera_manager_)
+        {
+            std::cerr << "creating manager" << std::endl;
+            camera_manager_ = std::make_shared<CameraManager>();
+            int ret = camera_manager_->start();
+            if (ret)
+                throw std::runtime_error("camera manager failed to start,"
+                                            "code " +
+                                            std::to_string(-ret));
+        }
+
+        return camera_manager_;
+    }
+
+    template <typename T>
+    class MessageQueue
+    {
+    public:
+        template <typename U>
+        void Post(U &&msg)
+        {
+            std::unique_lock<std::mutex> lock(mutex_);
+            queue_.push(std::forward<U>(msg));
+            cond_.notify_one();
+        }
+        T Wait()
+        {
+            std::unique_lock<std::mutex> lock(mutex_);
+            cond_.wait(lock, [this]
+                        { return !queue_.empty(); });
+            T msg = std::move(queue_.front());
+            queue_.pop();
+            return msg;
+        }
+        void Clear()
+        {
+            std::unique_lock<std::mutex> lock(mutex_);
+            queue_ = {};
+        }
+
+    private:
+        std::queue<T> queue_;
+        std::mutex mutex_;
+        std::condition_variable cond_;
+    };
+
+    void setupCapture();
+    void makeRequests();
+    void queueRequest(CompletedRequest *completed_request);
+    void requestComplete(Request *request);
+    void configureDenoise(const std::string &denoise_mode);
+
+    // std::unique_ptr<CameraManager> camera_manager_;
+    std::shared_ptr<Camera> camera_;
+    bool camera_acquired_ = false;
+    std::unique_ptr<CameraConfiguration> configuration_;
+    std::map<FrameBuffer *, std::vector<libcamera::Span<uint8_t>>> mapped_buffers_;
+    std::map<std::string, Stream *> streams_;
+    FrameBufferAllocator *allocator_ = nullptr;
```

Comment:
I propose to convert allocator from pointer to regular filed. No need to call new/delete for it each time. It allows to prevent memory leaks.